### ISSUE TITLE
Proper way of saving adversarial sound file

### ISSUE THREAD

**Issue by scfang** [Trusted-AI/adversarial-robustness-toolbox#967]
Hello,

I ran into an issue with saving and loading np.array sound examples generated from the Imperceptible_ASR attack class. Normally, I have been using torchaudio to write and load my audio file both original and adversarial. However, trying to pass the payload of the adversarial example after loading results in the example being classified correctly.

From what I examine, it seems that saving the file in a wav format causes the example to lose its adversarial properties. One way I could get around this is to use numpy.save to write the array to a file, then load the file so that none of the properties are lost. However, I would prefer to use wav files to test other classifiers with rather than numpy files.

Here is some rough code of how this can be recreated:

Generating and saving:
`speech_recognizer = PyTorchDeepSpeech(pretrained_model="librispeech")`
`asr_attack = ImperceptibleASRPyTorch( ... )`
`y = np.array([ ... ]) #target labels`
`x_adv = asr_attack.generate(source_sound, y)`
`adv_transcriptions = speech_recognizer.predict(x_adv, batch_size=len(x_adv), transcription_output=True)`
`for i in range(len(x_adv)):`
`file_loc = os.path.join("wav","name"+str(i)+".wav")`
`torchaudio.save(file_loc, torch.from_numpy(x_adv[i]), 16000)`

Loading and playing:
`sound, sample_rate = torchaudio.load(path)`
`sound = sound.squeeze().numpy()`
`transcription = speech_recognizer.predict(np.array([sound]), batch_size=len(audios), transcription_output=True)`

Here's an output of when I print out the np.array content of a freshly generated example and one loaded from disc:
`print(x_adv[0])`
`print(audios[0])`
`print(np.array_equal(audios[0], x_adv[0]))`
[-0.00227146 -0.00056028 -0.00029397 ... -0.00060096 -0.00071589
 -0.00074034]
[-0.00204474 -0.00015259  0.00094607 ... -0.00051881 -0.00079348
 -0.000824  ]
False

To summarize, I am wondering if it is possible to save an example with the properties preserved, specifically so I can test the examples on other classifiers. Any help is appreciated.

Regards

**Comment by beat-buesser** [comment]
Hi @scfang That's a great question.

@hesseltuinhof What do you think?

**Comment by hesseltuinhof** [comment]
I have been able to reproduce the issue, namely that using `torchaudio` to save and load Numpy audio samples does return sightly different audio samples. However, the reason for this behavior is better asked on the `torchaudio` repository. 

You can fix your issue by using Scipy`s `read` and `write` functions for WAV files. Saving and loading a WAV file will then return the exact same audio sample. Therefore I advise you to just use Scipy :relaxed:

The following script shows the difference of using `torchaudio` and `scipy`.

```python
import numpy as np
import torch
import torchaudio
from art.attacks.evasion import ImperceptibleASRPyTorch
from art.estimators.speech_recognition import PyTorchDeepSpeech
from scipy.io.wavfile import read, write

x = np.array([read("audio.wav")[1] / 2**15], dtype=np.float32) # Librispeech sample 3575-170457-0013.wav
y = np.array(["OLD WILL IS A FINE FELLOW BUT POOR AND HELPLESS SINCE MISSUS ROGERS HAD HER ACCIDENT"])

est = PyTorchDeepSpeech(pretrained_model="librispeech")
attack = ImperceptibleASRPyTorch(estimator=est, global_max_length=x.shape[1], max_iter_2=600)

x_adv = attack.generate(x, y)
y_adv = est.predict(x_adv, batch_size=len(x_adv), transcription_output=True)



torchaudio.save("audio_tmp.wav", torch.from_numpy(x_adv[0]), 16000)

x_adv_reload, _ = torchaudio.load("audio_tmp.wav")
x_adv_reload = x_adv_reload.squeeze().numpy()
y_adv_reload = est.predict(np.array([x_adv_reload]), batch_size=len(x_adv), transcription_output=True)

print("Torch save/load")
print(f"Proportion different: {(x_adv_reload != x_adv).sum() / x_adv.shape[1]}")

write("audio_tmp_2.wav", 16000, x_adv[0])

_, x_adv_reload = read("audio_tmp_2.wav")
y_adv_reload = est.predict(np.array([x_adv_reload]), batch_size=len(x_adv), transcription_output=True)

print("SciPy save/load")
print(f"Proportion different: {(x_adv_reload != x_adv).sum() / x_adv.shape[1]}")
```

It outputs:
```
Torch save/load
Proportion different: 0.9999455040871935
SciPy save/load
Proportion different: 0.0
```

**Comment by scfang** [comment]
@hesseltuinhof  @beat-buesser Thank you very much for spotting this! In hindsight I probably should have tried using different packages for saving/loading audio files. I will switch to using `scipy.io.wavfile` in the mean time.

Regards

**Comment by scfang** [comment]
UPDATE: 

It seems that using scipy.wave to read the wav file returns an nd.array type of np.int16, which messes things up when generating an example. I managed to renormalize the values as suggested by this stack post: 
[https://stackoverflow.com/a/42544738](https://stackoverflow.com/a/42544738)

Here is what it looks like in my code:
```   
from scipy.io.wavfile import read, write
  ...
def __getitem__(self, index):
        
       x_loc = self.audio_data.iloc[index][0]
       y_loc = self.audio_data.iloc[index][1]

       # Load data and get label
       _, x = read(x_loc)
       if type(x[0]) is np.int16:
           x =np.array([(num.astype(np.float32, order='K') / 32767.0) for num in x], dtype=np.float32)
       y, _ = self._parse_transcript(y_loc)
       return x, y
```
